### DISCUSSION TITLE
Provide nativescript-app-sync & server

### DISCUSSION THREAD

**Discussion by vallemar** [NativeScript/rfcs#35] [1. Ideas & Discussions]
A while ago I found https://github.com/NativeScript/nativescript-app-sync-server and this https://github.com/NativeScript/nativescript-app-sync. Try to get it to work under docker-compose and update my vue application. It cost me a lot to make it work on docker-compose and when I ran the application it did not update well. When I did it in a new app it worked but I understand that having a little more complexity and add-ons my application gave an error. I think it would be nice to provide the following:

- nativescript-app-sync-server: server to deploy to docker, docker-compose, kubernetes.
- nativescript-app-sync: solid and error-free update independent of flavor.

**Comment by asharghi** [comment]
I think this is really important for NS. Probably the single most important feature missing. Having the possibilty to update styling og components with app sync instead of going through Apple and Googles review prosess makes the development process  quicker, safer and easier. Not only is the review process time consuming but it takes some time before the user’s apps are automatically updated. Without app sync development is much slower because releases is much more «set in stone». Right noe I’m nervous every time I deploy a new version to the stores. It would be less stressful if patches and hotfixes where easy to get to the users phones. Really hope the NS team can look in to this.

**Reply by farfromrefug** [reply]
@asharghi i for one dont think i would ever use this. So saying it is the most important feature missing is weird t me.  But anyway, as it is a complete separated part of the N core you can already work on it and upgrade it to make it something solid for you and the others.
It would make a great addition to the N community ecosystem.

**Reply by asharghi** [reply]
I feel like we could have a long discussion on the importance of this. I will probably get back to you on that later, when I have more time to write. But I just wanted to reply to the part suggesting I can do it myself. Feel like that's a really unhelpful comment. If I had the technical knowledge and time I probably would, and the time needed for fixing this highly depends on the experience in the technology which I have none. Anyway, thanks for sharing your thoughts.

**Reply by philmday** [reply]
Just to like to add, I would like this capability. In a fast pace startup going after product-market fit, we don't have time to thoroughly test everything before releasing it. Being able to do hotfixes would be a huge win and would make us seriously consider doing a mobile app with nativescript. Until that point, we'll continue with our PWA approach.

**Reply by NathanWalker** [reply]
I agree this an important feature so thanks for raising this @vallemar @asharghi (cc/ @philmday ) - We are working on several infrastructure items including a revamp to Preview (single app with services to replace entire legacy playground) as well as other convenient services around it and this is on it. We are looking at providing a hosted app-sync server at some cost (given it's not free to host such a thing) however also providing it open source if others want to try and set one up themselves (just not a trivial task so probably not reasonable for every user to set one up on their own). Curious your thoughts on that when you get a chance.
/cc @rigor789

**Reply by asharghi** [reply]
Cordova and React Native developers can use Microsoft's CodePush (now moved to App Center) to deploy mobile app updates directly to their users devices. OTA updates are really easy this way. https://github.com/microsoft/appcenter-sdk-react-native. Some devs may want to pay for the hosted servers, but creating an App Center gets you a long way right now for free https://appcenter.ms/?utm_source=CodePush&utm_medium=Azure#pricing

My biggest argument for why I think this is important is because I find it challenging delivering a new version of an app quickly. Especially if a bug is detected that needs to be corrected immediately, failure to fix the bug as soon as possible can result in the loss of new customers, loss of sales, some financial loss, and negative customer experience. Even though the code can be fixed quickly, the new version of the app can take up to 1-4 business days for the appstore and playstore to review. And if it's approved it can take another week or so for all users to install the update. OTA skips all that and lets you live update it in a second.

If the new service for Preview can give production apps OTA updates then that will solve it, and App Center plugin won't be as necessary imo. I would pay for the hosted servers if it helped support NS, but having it opensource would be nice for people wanting to host it themselves  @nathanwalker

**Comment by CatchABus** [comment]
> A while ago I found https://github.com/NativeScript/nativescript-app-sync-server and this https://github.com/NativeScript/nativescript-app-sync. Try to get it to work under docker-compose and update my vue application. It cost me a lot to make it work on docker-compose and when I ran the application it did not update well. When I did it in a new app it worked but I understand that having a little more complexity and add-ons my application gave an error. I think it would be nice to provide the following:
> 
> * nativescript-app-sync-server: server to deploy to docker, docker-compose, kubernetes.
> * nativescript-app-sync: solid and error-free update independent of flavor.

What sort of issues did you face in docker? Perhaps, it's something that can be solved with a small patch.

**Reply by asharghi** [reply]
Maybe, but this is build upon the old MS CodePush. Hopefully we can create something using the new App Center. I think CodePush is a part of App Center so the technology underneath may be the same.

**Reply by edusperoni** [reply]
There were a couple of attempts trying to get NativeScript into Microsoft code push and subsequently app center, but the PRs and issues were closed.

Eventually they were open to the idea (discussed in an issue on the app center repo) but never replied again.

I'm on mobile now so it's difficult to find the issues, but you can search Microsoft's repos for NativeScript to find all the developments on this front.

We'd be happy to implement Microsoft's app center integration if they're willing to accept PRs and NativeScript into their ecosystem.

But as @NathanWalker said, we're trying to revamp the playground so the new "code push" might come through that implementation.

**Reply by asharghi** [reply]
Could we maybe create the repo under the NativeScript org and then get their attention afterwards. Or alternatively post the pr to ms and use the NS community on Discord to give it some attention with likes and comments so that ms team sees it.

Curious see how OTA works through the new preview :)

**Reply by vallemar** [reply]
finally I managed to make it work but with some errors, this is the repo and I think I am missing something to upload, I will not be home for a few days, when I can upload the last changes https://github.com/vallemar/nativescript-app-sync-server.  Although with this it worked when I updated from my application vue there were some errors, there were times that it was not updated, others the app was updated but when I killed the application and reopened it there was the previous version and other errors that I no longer remember.
